File: Top_100_Liked_Questions/581._Shortest_Unsorted_Continuous_Subarray/solution.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        umn = float('inf')
        umx = float('-inf')
        
        # find minimum and maximum unordered element
        for i in range(len(nums)-1):
            if nums[i] > nums[i+1]:
                umx = max(umx, nums[i])
                umn = min(umn, nums[i+1])
        
        # find the index where the minimum unordered element is supposed to be
        l,r = len(nums), -1
        for i in range(len(nums)):
            if nums[i] > umn:
                l = i
                break
                
        # find the index where the maximum unordered element is supposed to be
        for i in range(len(nums)-1,-1,-1):
            if nums[i] < umx:
                r = i
                break
        
        return max(r - l + 1, 0)
        
    
    def findUnsortedSubarray2(self, nums: List[int]) -> int:
        # naive solution - sort and compare
        sorted_nums = sorted(nums)
        l,r = len(nums), -1
        
        for i in range(len(nums)):
            if nums[i] != sorted_nums[i]:
                l = i
                break
        
        for i in range(len(nums)-1,-1,-1):
            if nums[i] != sorted_nums[i]:
                r = i
                break
        
        return max(r - l + 1, 0)

File: Top_100_Liked_Questions/581._Shortest_Unsorted_Continuous_Subarray/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findUnsortedSubarray = function(nums) {    
    // find the minimum and maximum unordered element
    let umin = Infinity;
    let umax = -Infinity;
    
    for (let i=0; i<nums.length-1; i++) {
        if (nums[i] > nums[i+1]) {
            umin = Math.min(umin, nums[i+1]);
            umax = Math.max(umax, nums[i]);
        }
    }
    
    // find the right index of the minimum unordered element
    let l = nums.length;
    for (let i=0; i<nums.length; i++) {
        if (nums[i] > umin) {
            l = i;
            break;
        }
    }
    
    // find the right index of the maximum unordered element
    let r = -1;
    for (let i=nums.length-1; i>-1; i--) {
        if (nums[i] < umax) {
            r = i;
            break;
        }
    }
    
    return Math.max(r - l + 1, 0);
}

var findUnsortedSubarray2 = function(nums) {
    // sort and find the first and last mismatch
    const sorted_nums = [...nums].sort((a,b) => a-b);
    
    let l = nums.length;
    let r = -1;
    
    for (let [i,v] of sorted_nums.entries()) {
        if (v !== nums[i]) {
            l = i;
            break;
        }
    }
    
    for (let i=nums.length-1; i>-1; i--) {
        if (sorted_nums[i] !== nums[i]) {
            r = i;
            break;
        }
    }
    
    return Math.max(r - l + 1, 0);
};